src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -348,107 +440,120 @@ internal enum BuildProgress
 
         private sealed class InProgressState
         {
-            private int _incrementDoNotAccessDirectly = 0;
-
             private readonly ExternalErrorDiagnosticUpdateSource _owner;
-            private readonly HashSet<ProjectId> _projectsBuilt = new HashSet<ProjectId>();
-            private readonly HashSet<ProjectId> _projectsWithErrors = new HashSet<ProjectId>();
 
-            private readonly Dictionary<ProjectId, ImmutableHashSet<string>> _allDiagnosticIdMap = new Dictionary<ProjectId, ImmutableHashSet<string>>();
-            private readonly Dictionary<ProjectId, ImmutableHashSet<string>> _liveDiagnosticIdMap = new Dictionary<ProjectId, ImmutableHashSet<string>>();
+            private readonly ConcurrentDictionary<ProjectId, ImmutableHashSet<string>> _allDiagnosticIdMap = new ConcurrentDictionary<ProjectId, ImmutableHashSet<string>>();
+            private readonly ConcurrentDictionary<ProjectId, ImmutableHashSet<string>> _liveDiagnosticIdMap = new ConcurrentDictionary<ProjectId, ImmutableHashSet<string>>();
 
+            // Fields that are used only from APIs invoked from serialized task queue, hence don't need to be thread safe.
```

Comment:
this comment is good.  but we need a comment on anything concurrent exaplining teh threading model around it.

Reply:
Yes, let me add a detailed comment. I myself tried to convert this whole type into thread safe one with locking throughout, and only later realized that the callers for majority of APIs on it are serialized.